flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/utils/testTableSources.scala

```diff
@@ -620,3 +622,39 @@ class TestInputFormatTableSource[T](
 
   override def getTableSchema: TableSchema = tableSchema
 }
+
+class TestDataTypeTableSource(
+    tableSchema: TableSchema,
+    values: Seq[Row]) extends InputFormatTableSource[Row] {
+
+  override def getInputFormat: InputFormat[Row, _ <: InputSplit] = {
+    new CollectionInputFormat[Row](
+      values.asJava,
+      fromDataTypeToTypeInfo(getProducedDataType)
+          .createSerializer(new ExecutionConfig)
+          .asInstanceOf[TypeSerializer[Row]])
+  }
+
+  override def getReturnType: TypeInformation[Row] =
+    throw new RuntimeException("Should not invoke this deprecated method.")
+
+  override def getProducedDataType: DataType = tableSchema.toRowDataType
+
+  override def getTableSchema: TableSchema = tableSchema
+}
+
+class TestStreamTableSource(
+    tableSchema: TableSchema,
+    values: Seq[Row]) extends StreamTableSource[Row] {
+
+  override def getDataStream(execEnv: StreamExecutionEnvironment): DataStream[Row] = {
+    execEnv.fromCollection(values, tableSchema.toRowType)
+  }
+
+  override def getReturnType: TypeInformation[Row] =
+    throw new RuntimeException("Should not invoke this deprecated method.")
```

Comment:
I would like to implement this method with returning `tableSchema.toRowType` instead of implementing `getProducedDataType`. Because currently, the behavior of `getProducedDataType` and return type of `getDataStream` is not align. For example, if users pass in a DECIMAL(5, 2), it might fail.

Reply:
ok